src/server/pkg/storage/fileset/index/reader.go

```diff
@@ -191,109 +196,45 @@ func (r *Reader) atEnd(name string) bool {
 	return name[:cmpSize] > r.filter.prefix[:cmpSize]
 }
 
-// Peek peeks ahead in the index.
-func (r *Reader) Peek() (*Header, error) {
-	if err := r.setupLevels(); err != nil {
-		return nil, err
-	}
-	if r.levels[0].peekHdr != nil {
-		return r.levels[0].peekHdr, nil
-	}
-	var err error
-	r.levels[0].peekHdr, err = r.Next()
-	return r.levels[0].peekHdr, err
-}
-
 func (r *Reader) callback(level int) chunk.ReaderFunc {
 	return func() ([]*chunk.DataRef, error) {
-		hdr, err := r.next(level - 1)
+		idx, err := r.next(level - 1)
 		if err != nil {
 			return nil, err
 		}
-		r.levels[level-1].currHdr = hdr
-		return hdr.Idx.DataOp.DataRefs, nil
+		return idx.DataOp.DataRefs, nil
 	}
 }
 
-// Copy is the basic data structure to represent a copy of data from
-// a reader to a writer.
-type Copy struct {
-	level int
-	raw   *chunk.Copy
-	hdrs  []*Header
-}
-
-// ReadCopyFunc returns a function for copying data from the reader.
-func (r *Reader) ReadCopyFunc(pathBound ...string) func() (*Copy, error) {
-	level := -1
-	var offset int64
-	var done bool
-	return func() (*Copy, error) {
-		if done {
-			return nil, io.EOF
-		}
-		// Setup levels, initialize first level.
-		if err := r.setupLevels(); err != nil {
-			return nil, err
-		}
-		if level < 0 {
-			level = len(r.levels) - 1
+func (r *Reader) Iterate(f func(*Index) error, pathBound ...string) error {
```

Comment:
I generally like to call this method `ForEach`, so then it reads as:

```go
r.ForEach(func (...) error {
    // Thing to do for each element.
})
```

I think it's slightly more standard, just based on grepping our vendor dir. The standard library has an `Iterate` method in the `reflect` package, but it doesn't take a cb, it returns an iterator struct.